Objects/unicodeobject.c

```diff
@@ -2163,12 +2163,27 @@ PyUnicode_FromStringAndSize(const char *u, Py_ssize_t size)
 PyObject *
 PyUnicode_FromString(const char *u)
 {
-    size_t size = strlen(u);
-    if (size > PY_SSIZE_T_MAX) {
-        PyErr_SetString(PyExc_OverflowError, "input too long");
-        return NULL;
+    /* Most calls of PyUnicode_FromString is for short ASCII string.
+     * So we optimize this function for them.
+     */
+    size_t len = 0;
+    int is_ascii = 1;
+
+    while (u[len] != '\0') {
+        if (u[len] > 127) {
+            is_ascii = 0;
+        }
+        len++;
+        if (len > PY_SSIZE_T_MAX) {
+            PyErr_SetString(PyExc_OverflowError, "input too long");
+            return NULL;
+        }
     }
-    return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)size, NULL, NULL);
+
+    if (is_ascii) {
+        return _PyUnicode_FromASCII(u, (Py_ssize_t)len);
+    }
+    return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)len, NULL, NULL);
```

Comment:
Why not modifying PyUnicode_DecodeUTF8Stateful() to detect if the input string is ASCII, or your heuristic is faster? Is it because it make non-ASCII string decoding slower?